### ISSUE TITLE
Apple7/8 (M1/M2): native Metal kernels, 2.1x decode and 2.3-2.7x prefill vs stock kernels on M1 Max

### ISSUE THREAD

**Issue by paperniuk** [incoai/splash#131]
## Summary

Splash now runs on the **M1 Max (Apple7) with native Metal kernels**. Compared with the stock Splash kernels on the same machine, decode is about **2.1x faster** for one request (with and without reasoning) and **3.6x faster** for four parallel requests, and prefill is **2.3-2.7x faster**. Output quality is unchanged within fp32 summation order.

This supersedes #95, whose Simdgroup corruption is fixed here.

Branch, rebased on current `main` (4 commits): https://github.com/paperniuk/splash/tree/apple7-m1-kernels

## What changed

Apple7/8 only; the Apple9+ kernel policy is untouched.

1. **Apple7 support** with a measured kernel policy (from #95).
2. **Fix for the Simdgroup decode corruption.** The K-split reduction read other threadgroups' partials through non-coherent L1. Marking `partials` as `coherent(device)` fixes it. This also changes the declarations in the Apple9 kernel; it follows the Metal spec, but I could not measure it on an M3. Plus an fp32-operand form (`SimdgroupF32`), since M1 has no bfloat arithmetic.
3. **`Mma64` prefill tile**: register-only simdgroup MMA with weights as exact half (`nibble | 0x6400` is 1024 + q) and activations as exact fp32.
4. **Decode verify kernel** with the same operands and one threadgroup for every lane of a batch, so each unpacked weight fragment feeds all lanes.

## Results

All measurements: M1 Max (32 GPU cores, 64 GB, macOS 26.7), Qwen3.8-27B, temperature 0, fresh server per run, ABBA order. "Before" is 8b76480 (the stock kernels with Apple7 enabled and a tuned policy).

**Splash's published serving methodology** (256 tokens, reasoning off, 32768 context limit, 64-token warmup, two repetitions per phase, medians, engine counter). The scenario prompt texts are not published, so these prompts are mine:

| Scenario | Before | After |
|---|---:|---:|
| technical | 16.6 tok/s | 34.3 tok/s |
| python_code | 18.0 tok/s | 37.4 tok/s |
| typescript_code | 19.5 tok/s | 40.4 tok/s |
| repeat | 35.9 tok/s | 75.7 tok/s |

**The five fixed prompts from @npanj's splash-plus benchmark** (https://github.com/npanj/splash-plus), run with their script unmodified (250 tokens, template default reasoning, i.e. xhigh): average 18.9 -> 39.1 tok/s (npanj reports 60.7 tok/s for stock Splash on an M5 Pro, 64 GB).

**Other conditions**: code with a 22-token prompt 26.2 -> 54.0 tok/s; reasoning low / medium / xhigh on a math problem 31.0 / 30.7 / 25.9 -> 66.5 / 67.2 / 54.6 tok/s; 8K context 12.8 -> 26.8 tok/s, 32K context 12.5 -> 23.8 tok/s; four parallel requests 18.0 -> 65.2 tok/s in total. Prefill: 2,048 tok 53 -> 142 tok/s, 32,029 tok 48 -> 110 tok/s.

**A real opencode session** (a coding agent rewriting a backend, reasoning medium, context growing from 30K to 89K tokens, one request at a time): decode 19.5-36.5 tok/s median per context range, prefill of new prompt tokens 96 tok/s at 16-32K and 54 tok/s at 64-96K, 95% of prompt tokens served from the prefix cache.

## Why the prefill tile

On the model's projection shapes at 2048 rows, the MPP tiles run at 2.4-3.5 TFLOPS on this GPU and `Mma64` at 7.3-7.4 (2 x rows x N x K / GPU time). A register-only MMA microbenchmark peaks at 8.6 TFLOPS for fp32 x fp32 and 9.9 for half x fp32; Apple rates the GPU at 10.4 TFLOPS FP32.

## Numerics and tests

Every q*x product is exact in fp32, and only the fp32 summation order changes. Against the previous M1 build's verify logits on 2528 shared positions: top-1 agreement 99.92%, mean KL 1.3e-4 nats, no bias in the chosen token's log-probability. On the rebased branch with macOS 26.7, `make test-engine` (168 PASS), vision parity and the model runtime oracle pass; the tests pin the Apple7/8 policy and the new dispatch grids.

M2 (Apple8) takes the same code path but is untested; relevant to #43.

Thanks for releasing Splash. Happy to adjust any of this to fit the development branch.

**Comment by JerilynFranz** [comment]
I'm getting failures on an M2 Ultra (c60) 128GB machine at 64K context:

```
% uvx llama-benchy --base-url http://localhost:8000/v1 --pp 65536 --tg 2048 --tokenizer ~/.omlx/models/scottlowry/Qwen3.8-27B-oQ8e-fp16-mtp/tokenizer.json --runs 1
No model specified, attempting to auto-detect from endpoint...
Auto-detected HF model: incoai/Qwen3.8-27B-Splash (served as: incoai/Qwen3.8-27B-Splash)
llama-benchy (0.4.0)
Date: 2026-09-24 10:52:40
Benchmarking model: incoai/Qwen3.8-27B-Splash at http://localhost:8000/v1
Concurrency levels: [1]
Loading text from cache: /Users/snowhare/.cache/llama-benchy/cc6a0b5782734ee3b9069aa3b64cc62c.txt
Total tokens available in text corpus: 144480
Warming up...
Warmup (User only) complete. Delta: 51 tokens (Server: 72, Local: 21)
Warmup (System+Probe) complete. Delta: 52 tokens (Server: 74, Local context: 21, Probe: 1)

Running coherence test...
Coherence test PASSED.
Measuring latency using mode: api...
Average latency (api): 1.22 ms
Running test: pp=65536, tg=2048, depth=0, concurrency=1
  Warmup (batch size 1)...
  Run 1/1 (batch size 1)...
HTTP 503: {"error":{"message":"engine is recovering; retry shortly","type":"server_error","code":"engine_recovering"}}
Printing results in MD format:



No results collected. Check if the model is generating tokens.
```

The log output shows

```

11:09:25 Error · not_found · GET /api/status
11:09:27 Error · not_found · GET /api/status
11:09:29 Error · not_found · GET /api/status
11:09:31 Error · not_found · GET /api/status
error: native transport stopped after an engine failure
11:09:32 Error · runtime_unavailable
11:09:32 Error · runtime_unavailable · POST /v1/chat/completions
11:09:32 Kernel policy for GPU family 8 with 60 cores.
```

I tried forcing --max-memory to 106G to see if memory auto-detection was failing but it didn't make any difference.

It works fine at 32K context.

**Comment by mer0vingi** [comment]
I ran these benchmarks with this branch on an M1 Ultra, 128 GB.

**Setup**
- Apple M1 Ultra, 48 GPU cores, 128 GB unified memory, macOS 26.6.1, Xcode 27.0
- Branch `apple7-m1-kernels` at `d2f902e`, built with `make`; the server logs `Kernel policy for GPU family 7 with 48 cores`
- Model `incoai/Qwen3.8-27B-Splash`, snapshot `9d27070b`
- `splash serve --max-context 32K`
- Following the methodology in this issue: 256 tokens, temperature 0, 64-token warmup, two repetitions, medians, engine counters (`timings` in the response). The prompts are my own.

**Decode, reasoning off**

| Scenario | tok/s |
|---|---:|
| technical | 50.0 |
| python_code | 64.7 |
| typescript_code | 43.6 |
| repeat | 98.3 |

**Reasoning effort (math problem)**

| Effort | tok/s |
|---|---:|
| low | 63.8 |
| medium | 65.5 |
| xhigh | 67.0 |

**Prefill and decode at depth** (unique prompts, no cache hits)

| Prompt | Prefill | TTFT | Decode |
|---|---:|---:|---:|
| 2,018 tok | 202 tok/s | 10.0 s | 31.6 tok/s |
| 8,176 tok | 197 tok/s | 41.4 s | 33.2 tok/s |
| 31,954 tok | 161 tok/s | 198 s | 21.2 tok/s |
| 31,954 tok, cached prefix | — | 0.46 s | — |

**Parallel requests** (aggregate tok/s, wall clock)

| Requests | tok/s |
|---|---:|
| 1 | 48.2 |
| 2 | 65.7 |
| 4 | 73.5 |

Output was coherent throughout. Thanks for the work on this branch!
